lib/sources.nix

```diff
@@ -270,6 +270,19 @@ let
       outPath = builtins.path { inherit filter name; path = origSrc; };
     };
 
+  predicateFilter = { name ? "source", src, predicate }:
+    let
+      isFiltered = src ? _isLibCleanSourceWith;
+      origSrc = if isFiltered then src.origSrc else src;
+      removeBase = lib.path.subpath.removePrefix (lib.path.deconstructPath origSrc).subpath;
+    in lib.cleanSourceWith {
+      inherit name src;
+      filter = absolutePath: type:
+        assert lib.isString absolutePath;
+        let subpath = removeBase (lib.substring 1 (-1) absolutePath);
+        in predicate { inherit type subpath; };
```

Comment:
This alternate implementation of the filter function blocks composability with the other functions. Both styles of filter function can be supported across the api, by extending the normalized representation that is returned by `toSourceAttributes`. Instead of the current `filter` attribute, it could have three attributes: `absFilter`, `relFilter` and `preferRelFilter`.